src/backend/drm/mod.rs

```diff
@@ -113,6 +113,16 @@ fn warn_legacy_fb_export() {
     });
 }
 
+/// Hint for DRM backend on how the surface should be presented
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PresentationMode {
+    /// Vertical synchronization
+    VSync,
+    /// Thearing presentationres
+    ASync,
+    // NOTE: Auto for detecting when thearing is desired would probably go here
+}
+
```

Comment:
We might be able to get away without introducing some `Auto` variant. After all every frame needs to have a preference anyway. By adding `PresentationMode` to `queue_frame` this can be nicely communicated (nice work btw :+1: ) frame by frame. I think it would be nice to keep it this way and always let downstream decide without some internal mechanism to detect it. What we should do instead is to provide some helper representing some sane behavior that can be used when downstream has no preference on its own.

I was thinking about doing it this way:
- Moving the `PresentationMode` enum to our render element module and introducing `RenderElement::presentation_mode(&self) -> PresentationMode` function
- Adding a helper to `RenderFrameResult` of `DrmCompositor` to get a sane default `PresentationMode`

The `RenderFrameResult` already has a reference to all `Element`s assigned to planes, so it should be as easy
as to verify that all planes have something assigned that prefers tearing.

This way downstream should have to complete freedom to decide, but can opt in to auto easily by just forwarding the `PresentationMode` from `RenderFrameResult` to `DrmCompositor::queue_frame`.

Reply:
Sounds good to me.

> The RenderFrameResult already has a reference to all Elements assigned to planes, so it should be as easy
as to verify that all planes have something assigned that prefers tearing.

This is a the part however that might warrent allowing customization in some way (perhaps with a closure, like we have the `default_primary_output_compare` and the likes?). E.g. some compositors might want to tear if a single element is added to the primary-plane that preferrs tearing (perhaps even dependant on the content-type?). To name a more concrete example, some FPS overlay on an overlay plane shouldn't be blocking tearing of an directly scanned-out fullscreen game on the primary plane. Of course interactions here can be extremly subtle, which is why I would also prefer to just delegate those issues to downstream.

Reply:
I mean, everything the helper function uses internally should be public accessible. So I don't think we need to complicate the default function, especially as it should be trivial to implement. The FPS overlay is a good example, so maybe just a single parameter if it should only check the primary plane or all assigned planes. Everything more complicated than that might be better handled completely downstream.

Reply:
Sounds good to me

Reply:
Done, but on `Element` rather than `RenderElement`, as `R: Renderer` generic was kinda annoying to deal with.

Elements can choose to request VSync, Async, or set no preference. Based on that we search for common denominator.

The whole change is self-contained in this new commit: https://github.com/Smithay/smithay/pull/1335/commits/a7e1c32fdb9f3dd1e94c9f106cacac5e8c678d7a